File: src/main/java/com/sau/hadoopweb/configuration/ExpenseHandler.java
package com.sau.hadoopweb.configuration;

import org.apache.spark.sql.Dataset;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.SparkSession;
import org.apache.spark.sql.functions;
import org.apache.spark.sql.types.Metadata;
import org.apache.spark.sql.types.StructType;
import org.apache.spark.sql.types.StructField;
import org.apache.spark.sql.types.DataTypes;

public class ExpenseHandler {

    public static void main(String[] args) throws Exception {
        // Initialize Spark session
        SparkSession spark = SparkSession.builder()
                .appName("Expense Handler")
                .master("local[*]") // Use local for testing. Replace with your cluster URL in production
                .config("spark.driver.extraJavaOptions", "--add-opens java.base/sun.nio.ch=ALL-UNNAMED")
                .config("spark.executor.extraJavaOptions", "--add-opens java.base/sun.nio.ch=ALL-UNNAMED")
                .config("spark.cassandra.connection.host", "localhost")  // Replace with your Cassandra host
                .getOrCreate();

        // Define the schema of the JSON message sent by the producer
        StructType expenseSchema = new StructType(new StructField[] {
                new StructField("userId", DataTypes.IntegerType, true, Metadata.empty()),
                new StructField("dateTime", DataTypes.StringType, true, Metadata.empty()),
                new StructField("description", DataTypes.StringType, true, Metadata.empty()),
                new StructField("type", DataTypes.StringType, true, Metadata.empty()),
                new StructField("count", DataTypes.IntegerType, true, Metadata.empty()),
                new StructField("payment", DataTypes.FloatType, true, Metadata.empty())
        });

        // Read data from Kafka topic
        Dataset<Row> rawDF = spark.readStream()
                .format("kafka")
                .option("kafka.bootstrap.servers", "localhost:9092")  // Kafka server URL
                .option("subscribePattern", "expense_*")  // Listen to all topics starting with "expense_"
                .load();

        // Convert Kafka 'value' column to String and filter out null messages
        Dataset<Row> messageDF = rawDF.selectExpr("CAST(value AS STRING) as message")
                .filter("message IS NOT NULL");

        // Parse JSON data into a structured DataFrame
        Dataset<Row> parsedDF = messageDF.select(functions.from_json(functions.col("message"), expenseSchema).alias("expense"));

        // Check if 'expense' column is not null before processing
        Dataset<Row> expenseDataDF = parsedDF.filter("expense.userId IS NOT NULL")
                .select("expense.userId", "expense.dateTime", "expense.description", "expense.type", "expense.count", "expense.payment");

        // Calculate the total expenses for each userId
        Dataset<Row> totalExpensesDF = expenseDataDF.groupBy("userId")
                .agg(functions.sum("payment").alias("total_expenses"));
        totalExpensesDF.printSchema();

        // Write the result to Cassandra
        totalExpensesDF.writeStream()
                .format("org.apache.spark.sql.cassandra")
                .option("checkpointLocation", "/tmp/spark_checkpoint")
                .option("keyspace", "expense")
                .option("table", "user_expenses")
                .outputMode("update")  // Use update mode for streaming aggregation
                .start()
                .awaitTermination();
    }
}

File: src/main/java/com/sau/hadoopweb/ExpenseHandler/src/main/scala/ExpenseHandler.scala
package com.sau.hadoopweb.ExpenseHandler.src.main.scala

import org.apache.spark.sql.{SparkSession, DataFrame, functions => F}
import org.apache.spark.sql.types._
import org.apache.spark.sql.streaming.Trigger
import org.apache.spark.sql.cassandra._
import com.datastax.spark.connector._

object ExpenseHandler {

  case class Expense(userId: Int, dateTime: String, description: String, typeExpense: String, count: Int, payment: Double)

  def main(args: Array[String]): Unit = {
    val spark = SparkSession
      .builder()
      .appName("Expense Handler")
      .config("spark.cassandra.connection.host", "localhost")
      .getOrCreate()

    import spark.implicits._

    val inputDF = spark
      .readStream
      .format("kafka")
      .option("kafka.bootstrap.servers", "localhost:9092")
      .option("subscribePattern", "expense_*")
      .load()

    val rawDF = inputDF.selectExpr("CAST(value AS STRING)").as[String]
    
    val expandedDF = rawDF.map(row => row.split(","))
      .map(row => Expense(
        row(0).toInt,
        row(1),
        row(2),
        row(3),
        row(4).toInt,
        row(5).toDouble
      ))

    val validExpensesDF = expandedDF.filter(expense => expense.payment > 0)

    val totalExpensesDF = validExpensesDF
      .groupBy("userId")
      .agg(F.sum("payment").alias("total_expenses"))

    val finalDF = totalExpensesDF.select("userId", "total_expenses")

    val query = finalDF.writeStream
      .trigger(Trigger.ProcessingTime("5 seconds"))
      .foreachBatch { (batchDF: DataFrame, batchID: Long) =>
        println(s"Writing to Cassandra $batchID")
        batchDF.write
          .cassandraFormat("user_expenses", "expense")
          .mode("append")
          .save()
      }
      .outputMode("update")
      .start()

    query.awaitTermination()
  }
}

